### ISSUE TITLE
Omit labels in record pattern matching

### ISSUE THREAD

**Issue by user471** [purescript/purescript#1690]
It would be nice if you could instead of 
```
case { a: 1, b: 2 } of
   { a:1, b:2 } -> "text"
```
do this
```
case { a: 1, b: 2 } of
   { 1, 2 } -> "text"
```

**Comment by jdegoes** [comment]
Puns for deconstruction make sense since we have them for construction.

**Comment by zudov** [comment]
Actually, we already have *exactly* that, and it works:

```purescript
> import Prelude
> case { a: 1, b: 2 } of { a, b } -> a + b
3
```
Note that it doesn't rely on the order of fields in the record, but only on their names.

**Comment by paf31** [comment]
Closing, this already exists.

**Comment by user471** [comment]
I fix the example, could you reopen it?

**Comment by paf31** [comment]
Ok, but bear in mind this syntax overlaps with that for record puns, so there would have to be a really good case for this.

**Comment by paf31** [comment]
As I mentioned elsewhere, this makes the meaning of a pattern context-sensitive.

**Comment by zudov** [comment]
Answering to https://github.com/purescript/purescript/issues/1687#issuecomment-162238918

`{ 1, "foo" }` indeed doesn't have a type, since it's a pattern/binder, however one could say that patterns are typed by the type of values that they match. For example pattern `{ foo = Just a }` only can match the values of type `forall a. { foo :: Maybe a }`, the record pun pattern `{ foo }` only matches values of type `forall a. { foo :: a }`.

Now let me rephrase @garyb's question. What is the type of values that `{ 1, "foo" }` matches? As @paf31 said, the meaning of such pattern is context-sensitive.

**Comment by natefaubion** [comment]
Is the issue just that we need syntax for matching multiple items at the same time without boxing them first? Could you not just have comma-separated expressions in the head of a case and in the patterns (without brackets of any kind). This is unambiguous grammatically, really useful, and doesn't require special record syntax.

**Comment by garyb** [comment]
That solves a use case of #1687 too. I don't think it would be too hard to add, as it's already how cases are represented in the compiler.

**Comment by user471** [comment]
>Now let me rephrase @garyb's question. What is the type of values that { 1, "foo" } can match?

You can easily describe this type - any record with two fields and fields would have type Int and String in same order.
If you want something more formal, then the language should have [anonymous records](https://github.com/purescript/purescript/issues/1267) and type would be {Int, String}

**Comment by zudov** [comment]
>  any record with two fields and fields would have type Int and String in same order

You can't encode such description with the current type system, you would have to name the fields explicitly. 

I am not really sure about that, but from the error messages it appears that when compiler typechecks `case expr of pattern -> result` it first infers the type of `pattern` and then matches the type of `expr` against it.

**Comment by user471** [comment]
>I am not really sure about that, but from the error messages it appears that when compiler typechecks case expr of pattern -> result it first infers the type of pattern and then matches the type of expr against it.

Maybe it can use ```expr``` to infer the type of pattern.
```
case { a: 1, b: 2 } of
   { 1, 2 } -> "text"
```
Then type of ```{1,2}``` would be ```{a:Int, b:Int}```

**Comment by zudov** [comment]
@natefaubion @garyb 

I just thought again about comma-separated expressions and realized:

```purescript
case [a, b, c] of
  [1, 2, 3] -> ...
```

I am not sure how big is overhead on creating array and then matching it like that, but since nobody gonna use it like this with more than a few elements overhead would be tiny.

Is there something that I missed? I don't think it's worth adding a feature with it's own syntax and concept, worth it when there is a already a simple solution.

**Comment by garyb** [comment]
The only problem with using arrays is they can't (conveniently) be heterogeneously typed.

**Comment by zudov** [comment]
@garyb Oh, true. That's what I missed

**Comment by paf31** [comment]
Resolved by #1696
